source/create.ts

```diff
@@ -1,13 +1,48 @@
+import {Readable as ReadableStream} from 'stream';
 import * as errors from './errors';
-import asStream from './as-stream';
-import asPromise from './as-promise';
-import {normalizeArguments, preNormalizeArguments} from './normalize-arguments';
-import merge, {mergeOptions, mergeInstances} from './merge';
+import {Options, Defaults, Method, NormalizedOptions, Response, CancelableRequest} from './utils/types';
 import deepFreeze from './utils/deep-freeze';
+import merge, {mergeOptions, mergeInstances} from './merge';
+import asPromise from './as-promise';
+import asStream, {ProxyStream} from './as-stream';
+import {preNormalizeArguments, normalizeArguments} from './normalize-arguments';
+import {Hooks} from './known-hook-events';
+
+const getPromiseOrStream = (options: NormalizedOptions): ProxyStream | CancelableRequest<Response> => options.stream ? asStream(options) : asPromise(options);
+
+export type HTTPAlias = 'get' | 'post' | 'put' | 'patch' | 'head' | 'delete';
 
-const getPromiseOrStream = options => options.stream ? asStream(options) : asPromise(options);
+export type ReturnResponse = (url: URL | string | Options & { stream?: false }, options?: Options & { stream?: false }) => CancelableRequest<Response>;
+export type ReturnStream = (url: URL | string | Options & { stream: true }, options?: Options & { stream: true }) => ProxyStream;
 
-const aliases = [
+export interface Got extends Record<HTTPAlias, ReturnResponse> {
+	stream: GotStream;
+	defaults: Defaults | Readonly<Defaults>;
+	GotError: typeof errors.GotError;
+	CacheError: typeof errors.CacheError;
+	RequestError: typeof errors.RequestError;
+	ReadError: typeof errors.ReadError;
+	ParseError: typeof errors.ParseError;
+	HTTPError: typeof errors.HTTPError;
+	MaxRedirectsError: typeof errors.MaxRedirectsError;
+	UnsupportedProtocolError: typeof errors.UnsupportedProtocolError;
+	TimeoutError: typeof errors.TimeoutError;
+	CancelError: typeof errors.CancelError;
+
+	(url: URL | string | Partial<Options & { stream?: false }>, options?: Partial<Options & { stream?: false }>): CancelableRequest<Response>;
+	(url: URL | string | Partial<Options & { stream: true }>, options?: Partial<Options & { stream: true }>): ProxyStream;
+	(url: URL | string | Options, options?: Options): CancelableRequest<Response> | ProxyStream;
+	create(defaults: Partial<Defaults>): Got;
```

Comment:
Are you sure we need all the `Partial`'s here?

Reply:
Technically, we don't need Partial there since Defaults already has everything as `value | undefined`